### ISSUE TITLE
Can't use api_version parameter in webhookEndpoints.create

### ISSUE THREAD

**Issue by sjdaniels** [jcberquist/stripe-cfml#37]
There is a new optional argument 'api_version' here:
https://stripe.com/docs/api/webhook_endpoints/create

This is for creating a webhook endpoint locked to a specific version. It looks like in lib/parsers/headers.cfc we are marking parameters named 'api_version' as headers and is therefore added to the ignoreArgs variable and are not included in the POST body.

I'm not sure the ramifications of removing this value from the headerKeys struct to the rest of the codebase. But likewise I can't think of another workaround to allow this parameter to pass through to the post body either.

Any thoughts or ideas for a workaround?

**Comment by jcberquist** [comment]
Yeah, I see this is going to be an issue and I suspect fixing it is going to involve breaking changes. Perhaps we need a new major version? 

My first thought is that I am not going to be able to allow arguments that are going to end up in the headers to be able to be passed right alongside the other arguments. For this particular route one might want to specify an API version for the request and a different one for the webhook 😬. Maybe a refactor to only ever pass header arguments in a headers struct?

**Comment by sjdaniels** [comment]
I've never passed headers alongside other arguments. What was the use case for that? I specify the api_version header via a module setting: config.apiVersion.

As a workaround I have been experimenting with simply removing "api_version" from the ignoreArgs array after the headers get parsed out. This seems to work but I guess it's possible that if someone is passing api_version as a normal argument, and expecting it to be sent as a header, it would also end up in the request body which would cause an API error if the method being called did not support that param.

Unfortunately I am in need of a quick fix for this. Because I am using the Identity beta product, I've been told I have until May 1 to update my webhook endpoints with the specific version "2020-08-27;identity_beta=v4" or they will stop receiving Identity webhook events. They told me this this morning! LOL. The only way to specify this is via the API (Dashboard does not support specifying version yet). So I think I will deploy my workaround fork to production to deal with this and then revert back.

I think your refactor sounds like the correct approach, but again, I'm not grokking the use case for passing headers along side so I don't fully grasp the potential breakage. Happy to help code/test if you can guide me with what need to be done.

**Comment by jcberquist** [comment]
So if you look at the headers section: <https://github.com/jcberquist/stripe-cfml#headers>, you can see the that for some headers you might expect to be setting them on a per request basis: e.g. `idempotencyKey` and `stripeAccount`. I decided to treat all of the headers in the same way. I do agree that `apiVersion` seems like something you would typically expect to set globally. However, it is there now, so I need to come up with a plan going forward. And I think just about anything reasonable will be a breaking change.

Some possibilities are:
1) Don't allow header arguments to be passed alongside regular arguments, but only in a headers struct.
2) If a headers struct is present, treat all other arguments as regular parameters and pass them in the request body. This might not be a backwards incompatible change, but it would create complex behavior, so I am not currently inclined to it.
3) I could remove the `apiVersion` config setting in favor of just using `stripeVersion`, which would remove this particular collision. 

I am actually somewhat inclined to 3) above since using `apiVersion` for the `Stripe-Version` header is really just an alternate name for the `stripeVersion` setting/parameter, which is what actually matches the header name. Of course, this would break your current config, for instance. But it would allow me to keep the functionality of how arguments are passed the same for now, though it would leave open the possibility of future name collisions. If I do this I would have to release a new major version as it would probably break everyone's configuration.

**Comment by elpete** [comment]
Looking at the Node library, it looks like they use a second or third parameter for header information like [API Version](https://stripe.com/docs/api/versioning) or [Idempotent Requests](https://stripe.com/docs/api/idempotent_requests). We could adopt a similar approach in our library. 

The challenge comes with how to handle the parameter as it exists now. I think the cleanest approach is a new major version change. Only a few keys would be moving positions, and many users wouldn't need to change anything. It would suit this use case very well since @sjdaniels could upgrade right away while other users can take the time they need to upgrade.

**Comment by jcberquist** [comment]
Thanks @elpete! I definitely agree that this will be a major version change. 

If I try to imitate the Node library more closely, I think I have two options. Right now, as I note in the readme, all three of the following are supported:
```cfc
charge = stripe.charges.create(
    {amount: 2000, currency: 'usd', source: cardToken},
    {stripeAccount: 'abc', idempotencyKey: 'def'}
);
charge = stripe.charges.create(
    params = {amount: 2000, currency: 'usd', source: cardToken},
    headers = {stripeAccount: 'abc', idempotencyKey: 'def'}
);
charge = stripe.charges.create(
    amount = 2000,
    currency = 'usd',
    source = cardToken,
    stripeAccount = 'abc',
    idempotencyKey = 'def'
);
```
So we could strictly imitate the Node library, and not support the last option at all, or we could alternately support the following syntax:
```cfc
charge = stripe.charges.create(
    amount = 2000,
    currency = 'usd',
    source = cardToken,
    stripeAccount = 'abc',
    headers = {
        idempotencyKey: 'def'
    }
);
```
Then the only possible collision would be the `headers` argument name. We could alternately name it `stripeHeaders` or `requestHeaders`, or something else, to mitigate the possibility. What do you think? I worry that not supporting this syntax at all could result in the most work for users to update.

**Comment by elpete** [comment]
I'm not sure of the usage of that third pattern, but I think you remove that. What happens when a param named headers crops up? I think it is the cleanest way forward. As a side benefit, I think it will also make the code easier to parse for users. Finally, I haven't seen that third usage out in any code; I would wager that it is not very common. A couple hours of find and replace at most. That's my vote, benevolent dictator. 😉

**Comment by jcberquist** [comment]
Very interesting, I do in fact use that syntax in my own code 😁. Do you always pass in a struct for the parameters? I agree that it would  be the simplest approach, but I had wanted to respect the fact that we can use named arguments in CFML. In fact, if you look at the Python SDK from Stripe, you can see they do support an approach like this:
```python
intent = stripe.PaymentIntent.retrieve(
  "pi_1DoRpz2eZvKYlo2CdQQ82XqG",
  stripe_version="2020-08-27"
)
charge = stripe.Charge.create(
  amount=2000,
  currency="usd",
  description="My First Test Charge (created for API docs)",
  source="tok_visa", # obtained with Stripe.js
  idempotency_key='Sin4BkDSPjRvaIQj'
)
```
The headers are mixed right in with the body parameters. I want to think about this some more. I really appreciate your input, thanks again.

**Comment by sjdaniels** [comment]
Thanks guys. I think imitating the Node library would be my preferred approach, but since I never pass headers this way I'm obviously biased.

Yes, I always pass params as a struct. The only simple value I pass would be in methods that require an ID, like:

`
customer = stripe.customer.update(
    'cus_123456',
    { email:"something@dealstream.com", phone:"1234567" }
)
`

I do not recall ever using named arguments, but agree it would be good to retain support for that. I'm in agreement with making this a major version update. I'm also on board with changing the config name from apiVersion to stripeVersion here - that's a very simple update obviously and it might help prevent other problems in the future.

**Comment by jcberquist** [comment]
Thanks guys! I have decided for now that I want to preserve the option for named arguments in v3, and will wait to pull them out when I see it is necessary (so in v4, I guess 😀). My plan for now is to do the following: remove the option to specify `apiVersion` for the `Stripe-Version` header when using named arguments. This mimics what the Python library has where the global setting is `api_version` but the per request setting is `stripe_version`. `apiVersion` will continue to work when positional arguments are used, as in that case the params and headers are able to be distinguished.

I am also going to remove the ability to do the following:
```cfc
charge = stripe.charges.create(
    params = {amount: 2000, currency: 'usd', source: cardToken},
    headers = {stripeAccount: 'abc', idempotencyKey: 'def'}
);
```
I think that one should either pick positional or named arguments, and avoid this hybrid situation.

**Comment by jcberquist** [comment]
@sjdaniels I have published `v3.0.0` to ForgeBox. It should fix this issue. Please let me know how it goes.
